Add option and function according issue #11 :

harveyquta opened the pull request:
1. Option
    --file, -f [path] : set config file path
         default : /usr/share/read-margin-temp/config.json
    --loose, -l : If have this option,
              ignore a sensor failed in that zone group

2. When detect sensor Functional, WarningAlarm, CtriticalAlarm property fail,
   set target dbus value to NaN. If Functional property is false, set NaN to sensor value.

Signed-off-by: Harvey Wu <Harvey.Wu@quantatw.com>

harveyquta committed:
Add option and function according issue #11 :

1. Option
    --file, -f [path] : set config file path
         default : /usr/share/read-margin-temp/config.json
    --loose, -l : If have this option,
              ignore a sensor failed in that zone group

2. When detect sensor Functional, WarningAlarm, CtriticalAlarm property fail,
   set target dbus value to NaN. If Functional property is false, set NaN to sensor value.

Signed-off-by: Harvey Wu <Harvey.Wu@quantatw.com>

Krellan commented on main.cpp:
```diff
@@ -73,7 +83,28 @@ int main(int argc, char **argv)
+        {
+            spofEnabled = false;
+        }
+        else
```
Please also add --debug option, to set debugEnabled to true.

harveyquta replied:
Done.

Krellan commented on main.cpp:
```diff
@@ -73,7 +83,28 @@ int main(int argc, char **argv)
+        }
+        else
+        {
+            inValiedoption = true;
```
Spelling: invalidOption

harveyquta replied:
Done.

Krellan commented on util/util.cpp:
```diff
@@ -15,12 +15,16 @@
+bool debugEnabled = false;
+
+// Flag for sensor fail
+bool errorEncountered = false;
```
It isn't strictly necessary to make errorEncountered a global. A return of NaN from getSensorDbusTemp() will set errorEncountered locally, see the std::isfinite() check on line 402.

harveyquta replied:
Oh, it's my mistake, I will fix it.
But I have a question for errEncountered, if not found the sensor's service path(line 63), is it need to make fanfail?
For example, if user's memory slot is not full and all memory is in a zone in config.json, it will let fan to fanfail mode in current code.

Krellan replied:
Good question. That is a policy decision. The correct answer depends on the user's decision. In our application, yes, all sensors must be present, and it is correct to declare an error condition if a sensor is missing. However, another user could feel differently. Do you think we should make this a configurable option?

harveyquta replied:
I think we can make a option for this situation that maybe useful for other user or in development phase.
I will implement this. Thanks.

harveyquta replied:
Done.

Krellan commented on util/util.cpp:
```diff
@@ -66,10 +70,67 @@ double getSensorDbusTemp(std::string sensorDbusPath, bool unitMilli)
+                                                      service,
+                                                      sensorDbusPath))
+        {
+            dbus::SDBusPlus::setValueProperty(bus,
```
Interesting, does this overwrite the source sensor with NaN if functional is false? I thought read-margin-temp only wrote the values of target sensors it is configured to use as output (the computed worst-margin virtual sensors). Overwriting the source sensors as well would be a change in behavior. I don't use functional, but if you feel this is OK, then it's OK, but I just wanted to point this out.

harveyquta replied:
Because when functional is false, the value property will not update and keep the value in last update.
The dbus-sensor will set [NaN](https://github.com/openbmc/dbus-sensors/blob/master/include/sensor.hpp#L344) to value property. But in phosphor-hwmon, it just use [continue](https://github.com/openbmc/phosphor-hwmon/blob/master/mainloop.cpp#L440) to skip the value update. That's why I want to set NaN to sensor value property in this place.
How do you think? Do I need to add this function when funtional is false?

Krellan replied:
That is interesting. Our application does not use phosphor-hwmon, it uses dbus-sensors exclusively. It almost seems like a bug in phosphor-hwmon, that it does not clear the old stale sensor value to NaN when there is a sensor error. The behavior of phosphor-hwmon and dbus-sensors is different here, which seems like a bug to me.

A better approach might be to fix this bug in phosphor-hwmon, instead of trying to do it here in read-margin-temp. Or, make this a configurable option, so that users can disable the NaN behavior if their application can not handle NaN and they would prefer to have stale readings survive, instead of clearing them to NaN here in read-margin-temp.

harveyquta replied:
OK, I see. I will remove this function in this place.

harveyquta replied:
Done.

Krellan reviewed (commented):
Looks good, I just had a few comments.

harveyquta commented:
Thanks for point out some problem. I add some question below the comments.

harveyquta committed:
Add "debug", "ignore" option

debug: enable debug mode to print log
ignore: enable to ignore if sensor service is empty

Signed-off-by: Harvey Wu <Harvey.Wu@quantatw.com>

Krellan commented:
Thanks. It looks good to me now.
